lib/honeybadger.ex

```diff
@@ -132,73 +130,87 @@ defmodule Honeybadger do
   end
 
-  defmacro notify(exception) do
-    macro_notify(exception, {:%{}, [], []}, [])
+  def notify(exception, metadata \\ %{}, stacktrace \\ []) do
+    notice = Notice.new(exception,
+                        contextual_metadata(metadata),
+                        backtrace(stacktrace))
+
+    Client.send_notice(notice)
   end
 
-  defmacro notify(exception, metadata) do
-    macro_notify(exception, metadata, [])
+  def context do
+    (Process.get(@context) || %{}) |> Enum.into(Map.new)
   end
 
-  defmacro notify(exception, metadata, stacktrace) do
-    macro_notify(exception, metadata, stacktrace)
+  def context(keyword_or_map) do
+    Process.put(@context, Map.merge(context(), Enum.into(keyword_or_map, %{})))
+    context()
   end
 
-  defp macro_notify(exception, metadata, stacktrace) do
-    quote do
-      Task.start fn ->
-        Honeybadger.do_notify(unquote(exception), unquote(metadata), unquote(stacktrace))
-      end
+  @doc """
+  Fetch configuration specific to the :honeybadger application.
+
+  ## Example
+
+      Honeybadger.get_env(:exclude_envs)
+      #=> [:dev, :test]
+  """
+  @spec get_env(atom) :: any | no_return
+  def get_env(key) when is_atom(key) do
+    case Application.fetch_env(:honeybadger, key) do
+      {:ok, {:system, var}} when is_binary(var) ->
+        System.get_env(var) || raise ArgumentError, "system variable #{inspect(var)} is not set"
+      {:ok, value} ->
+        value
+      :error ->
+        raise ArgumentError, "the configuration parameter #{inspect(key)} is not set"
     end
   end
 
-  def do_notify(exception, metadata, []) do
-    {:current_stacktrace, stacktrace} = Process.info(self(), :current_stacktrace)
-    do_notify(exception, metadata, stacktrace)
-  end
+  @doc """
+  Fetch all configuration specific to the :honeybadger application.
 
-  def do_notify(exception, %{context: _} = metadata, stacktrace) do
-    client = Client.new
-    backtrace = Backtrace.from_stacktrace(stacktrace)
-    notice = Notice.new(exception, metadata, backtrace)
-    Client.send_notice(client, notice)
-  end
+  This resolves values the same way that `get_env/1` does, so it resolves
+  :system tuple variables correctly.
 
-  def do_notify(exception, metadata, stacktrace) do
-    metadata = %{context: metadata}
-    do_notify(exception, metadata, stacktrace)
-  end
+  ## Example
 
-  def context do
-    (Process.get(@context) || %{}) |> Enum.into(Map.new)
+      Honeybadger.get_all_env()
+      #=> [api_key: "12345", environment_name: "dev", ...]
+  """
+  @spec get_all_env() :: [{atom, any}]
+  def get_all_env do
+    for {key, _value} <- Application.get_all_env(:honeybadger) do
+      {key, get_env(key)}
+    end
   end
 
-  def context(keyword_or_map) do
-    Process.put(@context, Map.merge(context(), Enum.into(keyword_or_map, %{})))
-    context()
-  end
+  # Helpers
 
   defp default_config do
-     [api_key: System.get_env("HONEYBADGER_API_KEY"),
+     [api_key: {:system, "HONEYBADGER_API_KEY"},
```

Comment:
I think this is the source of the config issue. `{:system, "HONEYBADGER_API_KEY"}` and `{:system, "MIX_ENV"}` should be environment variables, but they are being interpreted literally.

Reply:
Of course! The `get_all_env/0` function is being called before the merge, which is backward. I'll rework that now.